Guild/Channel repository updates.

CoreyShupe opened the pull request:
- Separated guild / channel repos.
- Added more events to each watching for most updates to items.
- Added more persistence stuff.
- Added tests to test every event.

CoreyShupe committed:
Guild/Channel repository updates.
- Separated guild / channel repos.
- Added more events to each watching for most updates to items.
- Added more persistence stuff.
- Added tests to test every event.

CoreyShupe committed:
Not sure when this got changed.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/repository/ChannelRepository.java:
```diff
@@ -0,0 +1,30 @@
+import io.reactivex.functions.Predicate;
+
+public interface ChannelRepository {
+  Maybe<Channel> getChannel(@NonNull Snowflake id);
```
No need for `@NonNull` here. By convention we assume non null

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/repository/ChannelRepository.java:
```diff
@@ -0,0 +1,30 @@
+public interface ChannelRepository {
+  Maybe<Channel> getChannel(@NonNull Snowflake id);
+
+  Single<ImmutableMap<Snowflake, Channel>> getChannels();
```
Why do we return a Map here, rather than something like `Observable<Channel>`?

CoreyShupe replied:
we could do that but for convenience, I might add both implementations just for variety and usability.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/repository/ChannelRepository.java:
```diff
@@ -0,0 +1,30 @@
+import io.reactivex.annotations.Nullable;
+import io.reactivex.functions.Predicate;
+
+public interface ChannelRepository {
```
I'm hesitant to have some many methods in the Repository.

The Repositories we want to have multiple implementations (e.g,. we might have one that caches in Redis, we might have one that uses a DB, we might have one that fetches info via the routes), so keeping it smaller makes it easier to implement these.

Those methods that are useful, but can be implemented purely by calling other methods of the repository I think we can move to the convenience layer.

CoreyShupe replied:
Gotcha, I was hesitant as well but I kinda figured some people don't know the number values of channel types, so I setup category lookups to serve that.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/repository/event/ChannelsFromEvents.java:
```diff
@@ -0,0 +1,207 @@
+  public ChannelsFromEvents(Observable<Event> events) {
+    // Merge all events which are meant to modify the channel by any means
+    channelWatcher =
+        events
```
This `events.ofType(clz).map(unwrap).map(doSeomthing)` pattern looks like a candidate for a method.

CoreyShupe replied:
Agreed.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/repository/event/ChannelsFromEvents.java:
```diff
@@ -0,0 +1,207 @@
+                    .map(ChannelUpdateEvent::unwrap)
+                    .map(UPDATE_CHANNEL_FUNCTION))
+            .scan(ImmutableMap.<Snowflake, Channel>of(), (map, function) -> function.apply(map))
+            .doOnError(Throwable::printStackTrace)
```
Don't use stdout

CoreyShupe replied:
Yeah, I need to do that, my bad.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/repository/event/GuildsFromEvents.java:
```diff
@@ -1,168 +1,275 @@
+            Role initial =
+                roles.stream().filter(role -> role.getId().equals(id)).findAny().orElse(null);
+            if (initial == null) {
+              throw new RuntimeException("Data not found where it should be.");
```
You could use `orElseThrow` instead of `orElse(null)`. I think we can do better than `RuntimeException`. Perhaps `IllegalStateException`?

CoreyShupe replied:
IllegalStateException works, I want to handle the exception myself instead of letting the optional throw.

ianagbip1oti commented on src/test/java/com/github/princesslana/eriscasper/faker/DataFaker.java:
```diff
@@ -78,6 +84,32 @@ public static UnavailableGuild unavailableGuildFromGuild(Snowflake predetermined
   }
 
+  public static Emoji emoji() {
+    return ImmutableEmoji.builder().addRoles(DiscordFaker.snowflake()).name(username()).build();
```
While `username` does what we want, it's not semantically clear, cos an emoji name is not a username

CoreyShupe replied:
hmm okay, I'll look into more string fakers.

ianagbip1oti commented on src/test/java/com/github/princesslana/eriscasper/faker/DataFaker.java:
```diff
@@ -78,6 +84,32 @@ public static UnavailableGuild unavailableGuildFromGuild(Snowflake predetermined
+    return ImmutableEmoji.builder().addRoles(DiscordFaker.snowflake()).name(username()).build();
+  }
+
+  public static GuildMember guildMember() {
```
Can't we add more randomness to these? e.g., a random boolean or long, rather than something constant.

CoreyShupe replied:
I'm sure we can, I'll look into more randominess

ianagbip1oti commented on src/test/java/com/github/princesslana/eriscasper/repository/event/TestChannelsFromEvents.java:
```diff
@@ -0,0 +1,177 @@
+  public void getChannel_whenGuildCreated_shouldCacheChannel() {
+    TestObserver<Channel> observer = new TestObserver<>();
+
+    // Create guild, define channel
```
If we're going to leave these comments in the tests let's stick with the semi-standard given, when, then.

Some of these comments are very close to just duplicating what the line of code says rather than adding info

CoreyShupe replied:
yeah I'm going to remove the comments, was just something for me to remember what I was doing. It's not necessary to have them at all.

CoreyShupe committed:
Addressed issues in repos.
- Useless @NonNull
- Trimming interfaces
- More randomness when faking
- Better faker for Emojis
- Pushed a pattern into a method
- Better exception usage
- Useless comment removal.

CoreyShupe committed:
Spotless
